### PR TITLE
[HttpFoundation] MongoDB driver support for sessions

### PR DESCRIPTION
**Pull request by funyx**
| Q | A |
| :-: | :-: |
| Branch? | 3.0 |
| Bug fix? | no |
| New feature? | Yes (MongoDB driver support) |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | N/A |

Adding support for the new MongoDB php driver in the mongodb sessions' handler. 

php-ref : [MongoDB driver](http://php.net/manual/en/set.mongodb.php)
mongo-ref : [official mongo-php library](http://mongodb.github.io/mongo-php-library/api/class-MongoDB.Collection.html)

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 1]
```diff
@@ -140,7 +140,7 @@ public function write($sessionId, $data)
             $this->options['expiry_field'] => $expiry,
         );
 
-        $this->getCollection()->update(
+        $this->getCollection()->updateOne(
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: This method does not always exist (see the failing test).

**Comment by xabbuh** [comment]
Status: Needs work

**Comment by funyx** [comment]
what am i doing wrong why cant i pass fabbot.io :/

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 2]
```diff
@@ -69,7 +69,7 @@ class MongoDbSessionHandler implements \SessionHandlerInterface
      */
     public function __construct($mongo, array $options)
     {
-        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo)) {
+        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo || $mongo instanceof \MongoDB\Client)) {
             throw new \InvalidArgumentException('MongoClient or Mongo instance required');
```
Comment (iltar) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Should add the new case here as well

Comment (funyx) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: $mongo instanceof \MongoDB\Client
-> MongoDB is available

any alternatives ?

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 3]
```diff
@@ -69,8 +69,10 @@ class MongoDbSessionHandler implements \SessionHandlerInterface
      */
     public function __construct($mongo, array $options)
     {
-        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo)) {
-            throw new \InvalidArgumentException('MongoClient or Mongo instance required');
+        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo || $mongo instanceof \MongoDB\Client)) {
+            throw new \InvalidArgumentException('MongoClient, Mongo or MongoDB instance required');
+        }elseif($mongo instanceof \MongoDB\Client && !class_exists('\MongoDB\Collection')){
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: `elseif` seems weird after throw an exception which is uncaught.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 4]
```diff
@@ -69,8 +69,10 @@ class MongoDbSessionHandler implements \SessionHandlerInterface
...
+            throw new \InvalidArgumentException('MongoDB/Collection instance not loaded');
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Wrong NS separator: `/` => `\`.

**Comment by phansys** [comment]
You need to respect CS in order to pass fabbot.io checks successfully. You can apply directly the patch proposed by the tool and push the changes again.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 5]
```diff
@@ -69,21 +69,24 @@ class MongoDbSessionHandler implements \SessionHandlerInterface
      */
     public function __construct($mongo, array $options)
     {
-        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo)) {
-            throw new \InvalidArgumentException('MongoClient or Mongo instance required');
+        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo || $mongo instanceof \MongoDB\Client)) {
+           throw new \InvalidArgumentException('MongoClient, Mongo or MongoDB instance required');
+        }elseif($mongo instanceof \MongoDB\Client && !class_exists('\MongoDB\Collection')){
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Please, put this check in a separated `if` statement.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 6]
```diff
@@ -108,9 +111,16 @@ public function close()
      */
     public function destroy($sessionId)
     {
-        $this->getCollection()->remove(array(
-            $this->options['id_field'] => $sessionId,
-        ));
+        if($this->getCollection() instanceof \MongoDB\Collection)
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Please, respect CS. You should take a look about [PSR-2](http://www.php-fig.org/psr/psr-2/#1-1-example) and [Symfony CS](http://symfony.com/doc/current/contributing/code/standards.html).

**Comment by phansys** [comment]
@xabbuh, do you know which version of PHP-CS-Fixer is fabbot.io using? I don't know how the current status of these changes could be passed the CS checks.

**Comment by xabbuh** [comment]
@phansys I think you have to ask @fabpot about that.

**Comment by funyx** [comment]
patched : running php-cs-fixer-1.11 locally

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 7]
```diff
@@ -80,10 +83,10 @@ public function __construct($mongo, array $options)
         $this->mongo = $mongo;
 
         $this->options = array_merge(array(
-            'id_field' => '_id',
-            'data_field' => 'data',
-            'time_field' => 'time',
-            'expiry_field' => 'expires_at',
```
Comment (GromNaN) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: The indentation should be let to 4 spaces.

**Comment by GromNaN** [comment]
Is it possible to add some test cases to [`MongoDbSessionHandlerTest`](https://github.com/symfony/symfony/blob/3.0/src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php) ?

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 8]
```diff
@@ -120,9 +129,15 @@ public function destroy($sessionId)
      */
     public function gc($maxlifetime)
     {
-        $this->getCollection()->remove(array(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: what about doing this (same below)?
```php
$collection = $this->getCollection();
$method = $collection instanceof \MongoDB\Collection ? 'deleteMany' : 'remove';
$collection->$method(array(
```

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 9]
```diff
@@ -69,8 +69,11 @@ class MongoDbSessionHandler implements \SessionHandlerInterface
      */
     public function __construct($mongo, array $options)
     {
-        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo)) {
-            throw new \InvalidArgumentException('MongoClient or Mongo instance required');
+        if (!($mongo instanceof \MongoClient || $mongo instanceof \Mongo || $mongo instanceof \MongoDB\Client)) {
+            throw new \InvalidArgumentException('MongoClient, Mongo or MongoDB instance required');
+        }
+        if ($mongo instanceof \MongoDB\Client && !class_exists('\MongoDB\Collection')) {
+            throw new \InvalidArgumentException('\MongoDB\Collection instance not loaded');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: can this really happen? it's possible that a `\MongoDB\Client` is instanciated, but `\MongoDB\Collection` isn't loadable?

**Comment by fabpot** [comment]
@funyx Can you finish this PR by taking comments into account? Thanks.

**Comment by fabpot** [comment]
Closing as there is no more feedback here and someone opened a similar one at #19186.

**Pull request closed without merging**